closure-size: move ncurses utilities to $out

ttuegel opened the pull request:
All the programs provided by ncurses were being installed to the $dev
output, but several of them are intended for runtime use, e.g. to
operate on the running terminal. These user-facing programs are moved to
the $bin output.

/cc @vcunat

vcunat commented:
Yes, I didn't know their real purpose when splitting. We'll also have to fixup the string references to these &ndash; IIRC there were a few already.

ttuegel commented:
Ah, thanks, I forgot to check for that!

vcunat commented:
Actually, maybe these binaries aren't worth splitting into another output. They use just 56 kB (by `du`) which is rather negligible wrt. > 5 MB of $out on which the libs depend (terminfo stuff).

vcunat commented:
I typically wasn't splitting such tiny outputs because each output brings some overhead &ndash; more inodes, entries in the database, larger risk of not being found, etc.

ttuegel commented:
That makes sense. I can modify this PR to do that instead; my only goal is getting them out of $dev.

vcunat commented on pkgs/development/libraries/ncurses/default.nix:
```diff
@@ -98,6 +98,13 @@ stdenv.mkDerivation rec {
+    moveToOutput "bin/clear" "$bin"
+    moveToOutput "bin/reset" "$bin"
+    moveToOutput "bin/tput" "$bin"
+    moveToOutput "bin/tset" "$bin"
```
Perhaps "bin/tabs" as well? I quickly looked over man pages for the binaries.

ttuegel commented:
@vcunat I agree about `bin/tabs`. What is the difference between `out` and `lib` currently? I get cyclic references when I put the binaries into `out`; can I combine both `bin` and `lib` into out?

vcunat commented:
The binaries depend on `lib` which depends on `out`. We should be able to put them to `lib` without any problems. I'm not sure if anything uses the terminfo data and not the libs &ndash; if not, it might be better to join those...

ttuegel commented:
The libraries are only 600K on top of the terminfo data's 6.5M; combining them gets us up to just under 7.2M. I don't see any reason not to do it.

ttuegel committed:
ncurses: combine $lib and $out outputs

The $lib output refers to the terminfo database in $out, which is about
10x larger than the ncurses shared library. Splitting these outputs
saves a small amount of space for any derivations that use the terminfo
database but not the ncurses library, but we do not have evidence that
any such exist.

ttuegel committed:
ncurses: move runtime utilities to $out

All the programs provided by ncurses were being installed to the $dev
output, but several of them are intended for runtime use, e.g. to
operate on the running terminal. These user-facing programs are moved to
the $bin output.

Several packages referred to "${ncurses}/bin" or "${ncurses.dev}/bin" at
runtime; these paths are also updated to refer to "${ncurses.bin}/bin".

ttuegel commented:
I am running my system with this now, with no ill effects.

ttuegel commented:
@vcunat As long as Hydra is so far behind on `closure-size`, could we merge this and cancel the earlier evaluation?

vcunat commented:
We should first extend this PR to fixup references from other packages &ndash; at least those explicit string references to bin/ of ncurses. Otherwise I would have merged it long ago.

vcunat commented:
BTW, no big hurry is needed because of Hydra, as it won't start building a significant amount of jobs from closure-size yet, as there are loads of other jobs with higher priority.

ttuegel commented:
Sorry, I didn't realize you were waiting on me! On it...

ttuegel commented:
OK, I think that should do it. I haven't tested everything, though--still rebuilding.

vcunat commented:
There's an evaluation problem due to removing `ncurses.lib`.

ttuegel commented:
Evaluation errors are fixed. I updated the PR title because it was slightly inaccurate :smile:
